README.md
# cool_nav

[![Pub Version (including pre-releases)](https://img.shields.io/pub/v/cool_nav?include_prereleases)](https://pub.dev/packages/cool_nav)
[![Github Action Status](https://github.com/masterashu/flutter_cool_nav/workflows/Dart%20CI/badge.svg)](https://github.com/masterashu/flutter_cool_nav/actions)

A collection of really awesome, easy to use BottomNavigationBars

## Getting Started

Add this to your package's pubspec.yaml file:

```yaml
---
dependencies:
  cool_nav: ^0.1.1
```

Now in your Dart code, you can use:

```dart
import 'package:cool_nav/cool_nav.dart';
```

## List of Navigation Bars:

#### Spotlight Navigation Bar

![Spotlight Navigation Bar](https://github.com/masterashu/flutter_cool_nav/blob/master/demo/spotlight_navigation_bar.gif?raw=true)  

> Based on [design](https://www.behance.net/gallery/94842819/Animated-Tab-Bar) made by [Sanchita Agarwal](https://www.linkedin.com/in/sanchita-agrawal-829a5612b).

An easy to use and customizable Navigation Bar. You can customize the
colors as well provide a custom [Gradient](https://api.flutter.dev/flutter/dart-ui/Gradient-class.html) for the spotlight.

**Usage**

```dart
Scaffold(
    // ...
    bottomNavigationBar: SpotlightNavigationBar(
        items: [
          SpotlightNavigationBarItem(icon: Icons.smartphone),
          SpotlightNavigationBarItem(icon: Icons.laptop_mac),
          SpotlightNavigationBarItem(icon: Icons.desktop_mac),
        ],
        selectedIndex: selectedIndex,
        selectedItemColor: Colors.cyan,
        onTap: _onTap,
    ),
)
```

#### Flip Box Navigation Bar

![Flip Box Navigation Bar](https://github.com/masterashu/flutter_cool_nav/blob/master/demo/flipbox_navigation_bar.gif?raw=true)  

> Based on [design](https://dribbble.com/shots/4811135-Tab-Bar-Cube-Interaction) made by [dannniel](https://dribbble.com/dannniel).

An easy to use and customizable Navigation Bar. You can customize the selected and unselected `Icons` and background `Colors` for the tiles.

**Usage**

```dart
Scaffold(
    // ...
    bottomNavigationBar: FlipBoxNavigationBar(
          selectedIndex: selectedIndex,
          verticalPadding: 20.0,
          items: <FlipBoxNavigationBarItem>[
            FlipBoxNavigationBarItem(
              name: 'Tasks',
              selectedIcon: Icons.done_all,
              selectedBackgroundColor: Colors.deepPurpleAccent[200],
              unselectedBackgroundColor: Colors.deepPurpleAccent[100],
            ),
            FlipBoxNavigationBarItem(
              name: 'People',
              selectedIcon: Icons.person,
              unselectedIcon: Icons.person_outline,
              selectedBackgroundColor: Colors.indigoAccent[200],
              unselectedBackgroundColor: Colors.indigoAccent[100],
            ),
            FlipBoxNavigationBarItem(
              name: 'Mail',
              selectedIcon: Icons.mail,
              unselectedIcon: Icons.mail_outline,
              selectedBackgroundColor: Colors.blueAccent[200],
              unselectedBackgroundColor: Colors.blueAccent[100],
            ),
          ]
    ),
)
```

Refer [here](https://github.com/masterashu/flutter_cool_nav/blob/master/example/lib/flipbox_nav_bar.dart) for an example.

More Awesome Navigation Bars will be added in the future.

## Examples

View the [example](https://github.com/masterashu/flutter_cool_nav/tree/master/example) folder to see more examples.


CHANGELOG.md
## [0.1.1] - 13-12-2023

Changes:
* [Maintenance by guyluz11](https://github.com/masterashu/flutter_cool_nav/pull/5)

## [0.1.0] - 13-05-2021

Changes:
* Migrated to Null Safety

## [0.0.3] - 29-04-2020

Changes:
* Added `FlipBoxBottomNavigationBar`.
* Divided diffrent files for examples.
* Added example for usage of `FlipBoxBottomNavigationBar`.

## [0.0.2] - 13-04-2020

Changes:
* Updated Package Information
* Added Docs for [SpotlightNavigationBarItem](https://pub.dev/documentation/cool_nav/latest/coolnav/SpotlightBottomNavigationBarItem-class.html)

## [0.0.1] - 13-04-2020

Initial Release.

Added `SpotlightNavigationBar`.


example/test/widget_test.dart
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('DST example Test 2', () {
    expect(1, 1);
  });
}


example/lib/spotlight_nav_bar.dart
import 'package:flutter/material.dart';
import 'package:cool_nav/cool_nav.dart';

class SpotlightNavBarHome extends StatefulWidget {
  SpotlightNavBarHome({Key? key}) : super(key: key);

  @override
  _SpotlightNavBarHomeState createState() => _SpotlightNavBarHomeState();
}

class _SpotlightNavBarHomeState extends State<SpotlightNavBarHome> {
  late int selectedIndex;

  _updateSelectedIndex(index) {
    setState(() {
      selectedIndex = index;
    });
  }

  @override
  void initState() {
    super.initState();
    selectedIndex = 0;
  }

  List<Text> texts = [
    Text("Page 1"),
    Text("Page 2"),
    Text("Page 3"),
    Text("Page 4")
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Spotlight"),
      ),
      body: Center(
        child: texts[selectedIndex],
      ),
      bottomNavigationBar: SpotlightNavigationBar(
        items: [
          SpotlightNavigationBarItem(icon: Icons.smartphone),
          SpotlightNavigationBarItem(icon: Icons.laptop_mac),
          SpotlightNavigationBarItem(icon: Icons.desktop_mac),
        ],
        selectedIndex: selectedIndex,
        selectedItemColor: Colors.cyan,
        onDestinationSelected: _updateSelectedIndex,
      ),
    );
  }
}


lib/src/flipbox_navbar/filpbox_navbar.dart
import 'dart:math' show pi, max;

import 'package:flutter/material.dart';

/// A Custom Navigation Bar that is displayed at the bottom of the
/// screen. The [FlipBoxNavigationBar] animated between the selected and
/// the unselected state it similar to a the rotation of a rectangle box.
/// This navigation bar is usually can be used in a [Scaffold], by providing
/// it as the [Scaffold.bottomNavigationBar] argument.
class FlipBoxNavigationBar extends StatefulWidget {
  /// Creates a flip box navigation bar which can be used with
  /// [Scaffold]'s [Scaffold.bottomNavigationBar] argument.
  ///
  /// The length of [items] should be at least three.
  ///
  /// The [selectedItemTheme], [unselectedItemTheme] and
  /// [backgroundColor] must not be null.
  ///
  /// The [verticalPadding] must be greater than 4.0.
  ///
  /// The [duration] can be used to have custom [Duration] of animation.
  /// [duration] must be greater than 100ms.
  FlipBoxNavigationBar({
    required this.items,
    this.selectedIndex = 0,
    this.selectedItemTheme =
        const IconThemeData(size: 24.0, color: Colors.black),
    this.unselectedItemTheme =
        const IconThemeData(size: 24.0, color: Colors.black),
    this.verticalPadding = 16,
    this.backgroundColor,
    this.textStyle,
    this.duration = const Duration(milliseconds: 800),
    this.onDestinationSelected,
    Key? key,
  })  : assert(items.length >= 3),
        assert(duration > const Duration(milliseconds: 100)),
        assert(verticalPadding >= 4.0),
        super(key: key);

  /// Defines the list of items to show on the navigation bar.
  final List<FlipBoxNavigationBarItem> items;

  /// The index of the selected [FlipBoxNavigationBar] in [items].
  final int selectedIndex;

  /// Called when one of the [items] is tapped.
  ///
  /// The stateful widget that creates the navigation bar needs to keep
  /// track of the index of the selected [FlipBoxNavigationBar] and call
  /// `setState` to rebuild the navigation bar with the new [selectedIndex].
  final ValueChanged<int>? onDestinationSelected;

  /// Defines the size and color of the item's icon when it is selected.
  final IconThemeData selectedItemTheme;

  /// Defines the size and color of the item's icon when it is unselected.
  final IconThemeData unselectedItemTheme;

  /// Defines the style of the item's name when it is selected.
  final TextStyle? textStyle;

  /// The background color of [FlipBoxNavigationBar] itself.
  final backgroundColor;

  /// The duration of animation.
  final Duration duration;

  /// The amount of padding to provide for each [FlipBoxNavigationBarTile].
  ///
  /// The height of tile is calculated by adding [verticalPadding] twice to the
  /// maximum icon size (selected / unselected).
  final double verticalPadding;

  /// Calculates the maximum size of selected/unselected icon size of [FlipBoxNavigationBarItem].
  double get maxIconSize {
    var iconSize = 24.0;
    iconSize = max(iconSize, selectedItemTheme.size ?? 24.0);
    iconSize = max(iconSize, unselectedItemTheme.size ?? 24.0);
    return iconSize;
  }

  @override
  _FlipBoxNavigationBarState createState() => _FlipBoxNavigationBarState();
}

/// An item that is used with [FlipBoxNavigationBar] containing the
/// information about the items to display.
class FlipBoxNavigationBarItem {
  /// Creates an item which is used with [FlipBoxNavigationBar.items].
  ///
  /// The [name], [selectedIcon] are required and must be non-null.
  ///
  /// The [selectedIcon] will be used if [unselectedIcon] is null.
  FlipBoxNavigationBarItem({
    required this.name,
    required this.selectedIcon,
    this.selectedBackgroundColor = Colors.blue,
    this.unselectedBackgroundColor = Colors.lightBlue,
    unselectedIcon,
  }) : unselectedIcon = unselectedIcon ?? selectedIcon;

  /// The text to be displayed for the selected item.
  final String name;

  /// The Icon which will be shown on the [FlipBoxNavigationBar] when the current
  /// item is selected.
  final IconData selectedIcon;

  /// The Icon which will be shown on the [FlipBoxNavigationBar] when the current
  /// item is unselected.
  final IconData unselectedIcon;

  /// The color of the [FlipBoxNavigationBarTile] when the current
  /// item is selected.
  final Color selectedBackgroundColor;

  /// The color of the [FlipBoxNavigationBarTile] when the current
  /// item is unselected.
  final Color unselectedBackgroundColor;
}

/// The Navigation Bar tile displayed within the [FlipBoxNavigationBar].
class FlipBoxNavigationBarTile extends StatefulWidget {
  /// Creates a new [FlipBoxNavigationBarTile] for displaying a single entry of an item
  /// from [FlipBoxNavigationBar.items].
  ///
  /// The [item] and [animation] is required.
  FlipBoxNavigationBarTile({
    required this.item,
    required Animation animation,
    this.height = 56.0,
    this.textStyle = const TextStyle(fontSize: 12.0),
    this.selectedIconTheme,
    this.unselectedIconTheme,
    Key? key,
  })  : assert(height >= 0.0),
        animation = _generateTween(animation),
        super(key: key);

  /// Defines the [FlipBoxNavigationBarItem] used to modify the UI of the tile.
  final FlipBoxNavigationBarItem item;

  /// An [Animation] object used to animate the transitions between selected
  /// and unselected state of Tile.
  final Animation animation;

  /// Defines the height of the [FlipBoxNavigationBarTile].
  final double height;

  /// Defines the [TextStyle] of the text shown with the selected Tile.
  final TextStyle textStyle;

  /// Defines the [IconThemeData] for the selected item.
  final IconThemeData? selectedIconTheme;

  /// Defines the [IconThemeData] for the unselected item.
  final IconThemeData? unselectedIconTheme;

  /// Interpolates an existing animation to a Bounce back [Curve] animation.
  static _generateTween(Animation animation) {
    return TweenSequence(<TweenSequenceItem>[
      TweenSequenceItem(tween: Tween(begin: 0.0, end: -0.1), weight: 10.0),
      TweenSequenceItem(tween: Tween(begin: -0.1, end: 0.0), weight: 25.0),
      TweenSequenceItem(tween: Tween(begin: 0.0, end: 1.0), weight: 20.0),
      TweenSequenceItem(tween: Tween(begin: 1.0, end: 1.1), weight: 30.0),
      TweenSequenceItem(tween: Tween(begin: 1.1, end: 1.0), weight: 20.0),
    ]).animate(animation as Animation<double>);
  }

  @override
  _FlipBoxNavigationBarTileState createState() =>
      _FlipBoxNavigationBarTileState();
}

class _FlipBoxNavigationBarState extends State<FlipBoxNavigationBar>
    with TickerProviderStateMixin {
  List<AnimationController>? _controllers;
  int? oldIndex;

  // Resets the controllers
  resetState() {
    // dispose any old controllers
    if (_controllers != null) {
      _controllers!.forEach((e) => e.dispose());
    }
    // Create new controllers
    _controllers = List<AnimationController>.generate(
      widget.items.length,
      (index) => AnimationController(
        duration: widget.duration,
        vsync: this,
      ),
    );
  }

  // Starts the flip animation
  switchIndex() {
    if (oldIndex != widget.selectedIndex) {
      _controllers![oldIndex!].reverse(from: 1.0);
      _controllers![widget.selectedIndex].forward(from: 0.0);
    } else {
      // partially animate the selected tile
      _controllers![widget.selectedIndex].forward(from: 0.8);
    }
  }

  @override
  void initState() {
    super.initState();
    resetState();
    oldIndex = widget.selectedIndex;
    switchIndex();
  }

  @override
  void didUpdateWidget(FlipBoxNavigationBar oldWidget) {
    super.didUpdateWidget(oldWidget);
    // Reset controllers when length of items is changed
    if (oldWidget.items.length != widget.items.length) {
      resetState();
      oldIndex = widget.selectedIndex;
    } else {
      oldIndex = oldWidget.selectedIndex;
    }
    switchIndex();
  }

  // build FlipBoxNavigationBarTile for all the items.
  _buildTiles() {
    var tiles = List<Widget>.generate(widget.items.length, (index) {
      return Expanded(
        child: GestureDetector(
          onTap: () {
            if (widget.onDestinationSelected != null) {
              widget.onDestinationSelected!(index);
            }
          },
          child: FlipBoxNavigationBarTile(
            item: widget.items[index],
            animation: _controllers![index],
            height: widget.verticalPadding * 2 + widget.maxIconSize,
            selectedIconTheme: widget.selectedItemTheme,
            unselectedIconTheme: widget.unselectedItemTheme,
            key: UniqueKey(),
          ),
        ),
      );
    });
    return tiles;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: MediaQuery.of(context).size.width,
      height: widget.verticalPadding * 2 + widget.maxIconSize,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceAround,
        children: _buildTiles(),
      ),
    );
  }

  @override
  void dispose() {
    _controllers!.forEach((e) => e.dispose());
    super.dispose();
  }
}

class _FlipBoxNavigationBarTileState extends State<FlipBoxNavigationBarTile> {
  // build selected size of the box
  get _selectedSide => Container(
        height: widget.height,
        color: widget.item.selectedBackgroundColor,
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: <Widget>[
            Icon(
              widget.item.selectedIcon,
              size: widget.selectedIconTheme?.size ?? 24.0,
              color: widget.selectedIconTheme?.color,
            ),
            Text(
              widget.item.name,
              textAlign: TextAlign.center,
              style: widget.textStyle,
            )
          ],
        ),
      );

  // build selected size of the box
  get _unselectedSide => Container(
        height: widget.height,
        color: widget.item.unselectedBackgroundColor,
        child: Center(
          child: Icon(
            widget.item.unselectedIcon,
            size: widget.unselectedIconTheme?.size,
            color: widget.unselectedIconTheme?.color,
          ),
        ),
      );

  @override
  Widget build(BuildContext context) {
    return Container(
      child: AnimatedBuilder(
        animation: widget.animation,
        builder: (_, child) {
          var selectedSide = Positioned(
            child: Transform.translate(
              offset: Offset(0.0, (1 - widget.animation.value) * widget.height),
              child: Transform(
                transform: Matrix4.identity()
                  ..setEntry(3, 2, 0.005)
                  ..rotateX((1 - widget.animation.value) * pi / 2)
                  ..scale(1 + (0.5 - _abs(widget.animation.value - 0.5)) / 5),
                alignment: Alignment.topCenter,
                child: _selectedSide,
              ),
            ),
          );
          var unselectedSide = Positioned(
            child: Transform.translate(
              offset: Offset(0.0, -widget.animation.value * widget.height),
              child: Transform(
                transform: Matrix4.identity()
                  ..setEntry(3, 2, 0.005)
                  ..rotateX(-widget.animation.value * pi / 2)
                  ..scale(1 + (0.5 - _abs(widget.animation.value - 0.5)) / 5),
                alignment: Alignment.bottomCenter,
                child: _unselectedSide,
              ),
            ),
          );
          return Stack(
            // Prevent selected size from painting over the unselected side
            // during the extra bounce animation.
            children: (widget.animation.value > 1.0)
                ? <Widget>[unselectedSide, selectedSide]
                : <Widget>[selectedSide, unselectedSide],
          );
        },
      ),
    );
  }
}

// Absolute value of a num
_abs(x) => x >= 0.0 ? x : -x;


lib/src/spotlight_navbar/spotlight_navbar.dart
import 'dart:math';

import 'package:flutter/material.dart';

/// A Custom Navigation Bar that is displayed at the bottom of the
/// screen. The [SpotlightNavigationBar] shows a spotlight over the
/// selected item. This navigation bar is usually can be used in a [Scaffold],
/// by providing it as the [Scaffold.bottomNavigationBar] argument.
class SpotlightNavigationBar extends StatefulWidget {
  /// Creates a spotlight navigation bar which can be used with
  /// [Scaffold]'s [Scaffold.bottomNavigationBar] argument.
  ///
  /// The length of [items] should be at least 2 and each item's icon property
  /// must not be null.
  ///
  /// The [selectedItemColor], [unselectedItemColor] and
  /// [backgroundColor] must not be null.
  ///
  /// The [iconSize] must be non-null and non-negative.
  ///
  /// If custom [IconThemeData] are used then it will override the
  /// [selectedItemColor] and [unselectedItemColor], and the custom
  /// [IconThemeData] must have a non-null [IconThemeData.color] and
  /// [IconThemeData.size].
  ///
  /// The [selectedItemColor] will be used for the Top Bar highlighting the
  /// active item regardless of [selectedIconTheme.color].
  ///
  /// The [spotlightGradient] parameter can be used to give a custom [Gradient]
  /// for the spotlight.
  SpotlightNavigationBar({
    Key? key,
    required this.items,
    this.onDestinationSelected,
    required this.selectedIndex,
    this.darkTheme = true,
    this.unselectedItemColor = const Color(0xff666666),
    this.selectedItemColor = const Color(0xffffffff),
    this.backgroundColor = const Color(0xff101010),
    this.iconSize = 24.0,
    this.unselectedIconTheme,
    this.selectedIconTheme,
    this.spotlightGradient,
  })  : assert(items.length >= 2),
        assert(0 <= selectedIndex && selectedIndex < items.length),
        assert(iconSize >= 0.0),
        assert(unselectedIconTheme == null ||
            (unselectedIconTheme.color != null &&
                unselectedIconTheme.size != null)),
        assert(selectedIconTheme == null ||
            (selectedIconTheme.color != null &&
                selectedIconTheme.size != null)),
        super(key: key);

  /// Defines the list of items which will be shown in the navigation bar.
  final List<SpotlightNavigationBarItem> items;

  /// Called when one of the [items] is tapped.
  ///
  /// The stateful widget that creates the navigation bar needs to keep
  /// track of the index of the selected [SpotlightNavigationBar] and call
  /// `setState` to rebuild the navigation bar with the new [selectedIndex].
  final ValueChanged<int>? onDestinationSelected;

  /// The index of the selected [SpotlightNavigationBarItem] in [items].
  final int selectedIndex;

  // TODO Implement Dark/Light Themes
  final bool darkTheme;

  /// The color of unselected [NavigationDestination.icon]s.
  final Color unselectedItemColor;

  /// The color of selected [NavigationDestination.icon], current item
  /// header and the color of the spotlight if [spotlightGradient] is not provided.
  final Color selectedItemColor;

  /// The color of the [SpotlightNavigationBar] itself.
  final Color backgroundColor;

  /// The size of the [SpotlightNavigationBarItem] icons.
  final double iconSize;

  /// The size, opacity, and color of the icon in the currently selected
  /// [SpotlightNavigationBarItem.icon].
  ///
  /// If this is not provided, the size will default to [iconSize], the color
  /// will default to [unselectedItemColor].
  ///
  /// It this field is provided, it must contain non-null [IconThemeData.size]
  /// and [IconThemeData.color] properties.
  final IconThemeData? unselectedIconTheme;

  /// The size, opacity, and color of the icon in the currently selected
  /// [SpotlightNavigationBarItem.icon].
  ///
  /// If this is not provided, the size will default to [iconSize], the color
  /// will default to [selectedItemColor].
  ///
  /// It this field is provided, it must contain non-null [IconThemeData.size]
  /// and [IconThemeData.color] properties.
  final IconThemeData? selectedIconTheme;

  /// The custom [Gradient] to use for the spotlight of the selected
  /// [NavigationDestination].
  final Gradient? spotlightGradient;

  @override
  _SpotlightNavigationBarState createState() => _SpotlightNavigationBarState();
}

class _SpotlightNavigationBarState extends State<SpotlightNavigationBar>
    with TickerProviderStateMixin {
  late AnimationController animation;

  // The previous selectedIndex of the widget. Used to animate the top bar.
  late int oldIndex;

  _resetState() {
    animation = AnimationController(
      duration: const Duration(milliseconds: 400),
      animationBehavior: AnimationBehavior.preserve,
      vsync: this,
    );
  }

  @override
  void initState() {
    super.initState();
    _resetState();
    oldIndex = widget.selectedIndex;
    // only show spotlight FadeIn Animation
    animation.forward(from: 0.5);
  }

  @override
  void didUpdateWidget(SpotlightNavigationBar oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.selectedIndex != oldWidget.selectedIndex) {
      _resetState();
      oldIndex = oldWidget.selectedIndex;
    }
    if (widget.items.length != oldWidget.items.length) {
      _resetState();
      oldIndex = widget.selectedIndex;
    }
    animation.forward();
  }

  @override
  void dispose() {
    animation.dispose();
    super.dispose();
  }

  IconThemeData _getThemeData(bool active) {
    if (active) {
      return widget.selectedIconTheme ??
          IconThemeData(color: widget.selectedItemColor, size: widget.iconSize);
    } else {
      return widget.unselectedIconTheme ??
          IconThemeData(
              color: widget.unselectedItemColor, size: widget.iconSize);
    }
  }

  _buildItems(Animation opacityTween) {
    List<Widget> tiles = [];
    for (var i = 0; i < widget.items.length; i++) {
      tiles.add(Expanded(
        child: GestureDetector(
          child: Center(
            child: Stack(
              children: [
                if (widget.selectedIndex == i &&
                    opacityTween.status == AnimationStatus.completed)
                  Opacity(
                    opacity: opacityTween.value,
                    child: CustomPaint(
                      painter: _SpotlightPainter(
                        offset: Offset(widget.iconSize / 2 + 3.5, 0),
                        iconSize: widget.iconSize,
                        gradient: widget.spotlightGradient,
                        color: widget.selectedIconTheme?.color ??
                            widget.selectedItemColor,
                      ),
                    ),
                  ),
                _SpotlightNavigationBarTile(
                  key: UniqueKey(),
                  icon: widget.items[i].icon,
                  iconTheme: _getThemeData(
                      (widget.selectedIndex == i) && animation.value >= 0.7),
                ),
              ],
            ),
          ),
          onTap: () {
            if (widget.onDestinationSelected != null) {
              widget.onDestinationSelected!(i);
            }
          },
          behavior: HitTestBehavior.translucent,
        ),
      ));
    }
    return tiles;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: max(56.0, (widget.iconSize + 24)),
      width: double.infinity,
      decoration: BoxDecoration(color: widget.backgroundColor),
      child: AnimatedBuilder(
        animation: animation,
        builder: (_, child) {
          Animation opacityTween = (animation.value >= 0.5)
              ? Tween(begin: 0.0, end: 1.0).animate(
                  CurvedAnimation(parent: animation, curve: Interval(0.8, 1.0)))
              : Tween(begin: 1.0, end: 0.0).animate(CurvedAnimation(
                  parent: animation, curve: Interval(0.0, 0.2)));

          return Stack(
            children: <Widget>[
              Row(
                children: _buildItems(opacityTween),
                mainAxisAlignment: MainAxisAlignment.spaceAround,
              ),
              CustomPaint(
                painter: _SelectedItemHeaderPainter(
                  iconSize: widget.iconSize,
                  count: widget.items.length,
                  color: widget.selectedIconTheme?.color ??
                      widget.selectedItemColor,
                  oldPosition: oldIndex,
                  newPosition: widget.selectedIndex,
                  animation: Tween(begin: 0.0, end: 1.0).animate(
                      CurvedAnimation(
                          parent: animation, curve: Interval(0.1, 0.6))),
                ),
                child: SizedBox(
                  width: double.infinity,
                ),
              ),
            ],
          );
        },
      ),
    );
  }
}

/// An item that is used with [SpotlightNavigationBar] containing the
/// information about the items to display.
class SpotlightNavigationBarItem {
  /// Creates an item that is used with [SpotlightNavigationBar.items].
  ///
  /// The argument [icon] should not be null when used in a Material Design's [NavigationBar].
  SpotlightNavigationBarItem({required this.icon});

  /// The Icon which will be shown on the [SpotlightNavigationBar]
  final IconData icon;
}

class _SelectedItemHeaderPainter extends CustomPainter {
  _SelectedItemHeaderPainter({
    required this.oldPosition,
    required this.newPosition,
    required this.animation,
    required this.count,
    this.iconSize = 24,
    this.color = Colors.white,
  });
  final int oldPosition, newPosition, count;
  final double iconSize;
  final Color color;
  final Animation animation;

  double get progress => animation.value;

  Offset startOffset(Size size) {
    var freeSpace = (size.width / count) - iconSize;
    assert(freeSpace >= 0);
    return Offset(
        freeSpace / 2 +
            (freeSpace + iconSize) *
                (oldPosition * (1 - progress) + newPosition * progress),
        1);
  }

  Offset endOffset(Size size) {
    var freeSpace = (size.width / count) - iconSize;
    assert(freeSpace >= 0);
    return Offset(
        freeSpace / 2 +
            (freeSpace + iconSize) *
                (oldPosition * (1 - progress) + newPosition * progress) +
            iconSize,
        1);
  }

  @override
  void paint(Canvas canvas, Size size) {
    var p = Paint()
      ..color = this.color
      ..strokeWidth = 2;
    canvas.drawLine(startOffset(size), endOffset(size), p);
  }

  @override
  bool shouldRepaint(_SelectedItemHeaderPainter oldDelegate) => true;
}

class _SpotlightPainter extends CustomPainter {
  final Offset offset;
  final double iconSize;
  final Gradient? gradient;
  final Color? color;

  _SpotlightPainter({
    this.iconSize = 24,
    this.offset = Offset.zero,
    this.gradient,
    this.color,
  });

  double get spotlightSize => max(60, iconSize + 36);

  @override
  void paint(Canvas canvas, Size size) {
    var rect = Rect.fromLTWH(0, 0, spotlightSize, spotlightSize);
    var p = Paint()
      ..shader = (gradient != null)
          ? gradient!.createShader(rect)
          : LinearGradient(colors: <Color>[
              color!.withAlpha(64),
              color!.withAlpha(36),
              color!.withAlpha(0)
            ], begin: Alignment.topCenter, end: Alignment.bottomCenter)
              .createShader(rect);
    var path = Path()..addPolygon(getPath(size), true);
    canvas.drawPath(path, p);
  }

  @override
  bool shouldRepaint(_SpotlightPainter oldDelegate) => true;

  List<Offset> getPath(Size size) {
    var points = <Offset>[];
    points.add(offset + Offset(2, 0));
    points
        .add(offset + Offset(iconSize / 2 - spotlightSize / 2, spotlightSize));
    points
        .add(offset + Offset(iconSize / 2 + spotlightSize / 2, spotlightSize));
    points.add(offset + Offset(iconSize - 2, 0));
    return points;
  }
}

class _SpotlightNavigationBarTile extends StatelessWidget {
  final IconData? icon;
  final IconThemeData? iconTheme;

  const _SpotlightNavigationBarTile({Key? key, this.icon, this.iconTheme})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      height: max(56, iconTheme!.size! + 24),
      width: max(56, iconTheme!.size! + 24),
      child: Center(
        child: IconTheme(
          child: Icon(icon),
          data: iconTheme!,
        ),
      ),
    );
  }
}


example/lib/flipbox_nav_bar.dart
import 'package:flutter/material.dart';
import 'package:cool_nav/cool_nav.dart';

class FlipBoxNavigationBarHome extends StatefulWidget {
  FlipBoxNavigationBarHome({Key? key}) : super(key: key);

  @override
  _FlipBoxNavigationBarHomeState createState() =>
      _FlipBoxNavigationBarHomeState();
}

class _FlipBoxNavigationBarHomeState extends State<FlipBoxNavigationBarHome> {
  late int selectedIndex;

  _updateSelectedIndex(index) {
    setState(() {
      selectedIndex = index;
    });
  }

  @override
  void initState() {
    super.initState();
    selectedIndex = 0;
  }

  List<Text> texts = [
    Text('Page 1'),
    Text('Page 2'),
    Text('Page 3'),
    Text('Page 4'),
    Text('Page 5')
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text('Flipbox'),
        ),
        body: Center(
          child: texts[selectedIndex],
        ),
        bottomNavigationBar: FlipBoxNavigationBar(
          selectedIndex: selectedIndex,
          verticalPadding: 20.0,
          items: <FlipBoxNavigationBarItem>[
            FlipBoxNavigationBarItem(
              name: 'Tasks',
              selectedIcon: Icons.done_all,
              unselectedIcon: Icons.done_outline,
              selectedBackgroundColor: Colors.deepPurpleAccent[200]!,
              unselectedBackgroundColor: Colors.deepPurpleAccent[100]!,
            ),
            FlipBoxNavigationBarItem(
              name: 'People',
              selectedIcon: Icons.person,
              unselectedIcon: Icons.person_outline,
              selectedBackgroundColor: Colors.indigoAccent[200]!,
              unselectedBackgroundColor: Colors.indigoAccent[100]!,
            ),
            FlipBoxNavigationBarItem(
              name: 'Mail',
              selectedIcon: Icons.mail,
              unselectedIcon: Icons.mail_outline,
              selectedBackgroundColor: Colors.blueAccent[200]!,
              unselectedBackgroundColor: Colors.blueAccent[100]!,
            ),
            FlipBoxNavigationBarItem(
              name: 'Flagged',
              selectedIcon: Icons.flag,
              unselectedIcon: Icons.outlined_flag,
              selectedBackgroundColor: Colors.greenAccent[200]!,
              unselectedBackgroundColor: Colors.greenAccent[100]!,
            ),
            FlipBoxNavigationBarItem(
              name: 'Accounts',
              selectedIcon: Icons.people,
              unselectedIcon: Icons.people_outline,
              selectedBackgroundColor: Colors.orangeAccent[200]!,
              unselectedBackgroundColor: Colors.orangeAccent[100]!,
            ),
          ],
          onDestinationSelected: _updateSelectedIndex,
        ));
  }
}
